docs/timeline.md

```diff
@@ -34,9 +31,35 @@ erDiagram
+
+## Details
+
+But, again, there's no relational integrity with any of this. So looking up a post's creator's followers to determine whether it should be excluded by this WHERE is a non-starter. This means we need to do some pre-processing of our data to make it easy to query without those relations. The solution to that is the `Audience` concept. Before they're added to the tsdb, posts get lightly pre-processed to generate audience keys, and these keys essentially constitute feeds, which can be easily merged at query time. Typical audiences would be `public`, the creator's `followers`, and the `mentions` for anyone mentioned in the post. Then to populate a feed, the timeline service will query the feeds table with something like this:
+
+```sql
+SELECT time_stamp, post_id, audience, boosted_by
+FROM feeds
+WHERE audience in ('set', 'of', 'audience', 'keys', 'for', 'the', 'feed')
+    AND time_stamp >= $start_time
+FETCH $limit
+OFFSET $start
+GROUP BY (post_id)
+ORDER BY time_stamp desc
+```
+
+The query engine is able to do this very very fast, because there are no foreign or unique keys or the related constraints on this table. And everything is already ordered by the timestamp.
+
+The common case is the home feed, where the set of audiences is the profile's mentions, follows, and follows' boosts. But this same process applies to other scenarios. Lists as well as filters for boosts and media are well established alternatives, but this mechanism would actually allow us to create feeds for anything where we can identify a useful audience key. In ActivityPub terms, just about any collection could have an audience. ActivityPub also specifies that ability for actors to define arbitrary collections for themselves. This would permit people to curate lists or categories of their own content, and Letterbook users would be able to follow those categories. Assuming the sending server informs us about the category, that is.
+
+If you noticed that this means we'll be duplicating a lot of data in this table, you get a gold star. This is again not a significant concern thanks to the properties of timeseries databases. There are two factors behind this. One, is that TSDBs can achieve extremely good at-rest compression by using column-oriented storage. Two, these databases are very good at ejecting old data. In fact, if you're fine with letting the db do whatever is easy (and I am), this happens basically for free. Everything is already time-ordered and stored as columns on disk. Those columns are already further partitioned by time, as a query optimization. Because of this, partitions that are older than your chosen expiration time can just be discarded with practically no effort.
```

Comment:
I'm not sure I understand the implications here. If things are automatically removed from the DB what happens to a users personal feed over time? Will they lose access to data? Also, a big cost in cloud-ops can be attributed to storage size, especially for a larger instance this may not be a trivial cost.

Reply:
That's a good point about a single user's timeline. It could be regenerated from the regular relational store, but that might not be desirable. We might need to exempt those feeds from expiring, or find other handling for it.

As far as the storage used by the TSDB, that should stay small. The compression really is bonkers good, and we can certainly eject duplicates after some time. For example, public posts will generally go to the followers and public audiences. We can expire the public duplicate. And that expiration window is easy to configure.

https://www.timescale.com/products#controlCosts